soroban-sdk/src/muxed_address.rs

```diff
@@ -0,0 +1,310 @@
+                )
+                .unwrap(),
+            ),
+        }
+    }
+
+    #[inline(always)]
+    pub(crate) unsafe fn unchecked_new_from_address(env: Env, obj: AddressObject) -> Self {
+        Self {
+            env,
+            obj: AddressObjectWrapper::Address(obj),
+        }
+    }
+
+    #[inline(always)]
+    pub(crate) unsafe fn unchecked_new(env: Env, obj: MuxedAddressObject) -> Self {
+        Self {
+            env,
+            obj: AddressObjectWrapper::MuxedAddress(obj),
+        }
+    }
+
+    #[inline(always)]
+    pub fn env(&self) -> &Env {
+        &self.env
+    }
+
+    pub fn as_val(&self) -> &Val {
+        match &self.obj {
+            AddressObjectWrapper::Address(o) => o.as_val(),
+            AddressObjectWrapper::MuxedAddress(o) => o.as_val(),
+        }
+    }
+
+    pub fn to_val(&self) -> Val {
+        match self.obj {
+            AddressObjectWrapper::Address(o) => o.to_val(),
+            AddressObjectWrapper::MuxedAddress(o) => o.to_val(),
+        }
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFromVal<Env, ScVal> for MuxedAddress {
+    type Error = ConversionError;
+    fn try_from_val(env: &Env, val: &ScVal) -> Result<Self, Self::Error> {
+        let v = Val::try_from_val(env, val)?;
+        match val {
+            ScVal::Address(sc_address) => match sc_address {
+                ScAddress::Account(_) | ScAddress::Contract(_) => {
+                    Ok(AddressObject::try_from_val(env, &v)?
+                        .try_into_val(env)
+                        .unwrap_infallible())
+                }
+                ScAddress::MuxedAccount(_) => Ok(MuxedAddressObject::try_from_val(env, &v)?
+                    .try_into_val(env)
+                    .unwrap_infallible()),
+                ScAddress::ClaimableBalance(_) | ScAddress::LiquidityPool(_) => {
+                    panic!("unsupported ScAddress type")
+                }
+            },
+            _ => panic!("incorrect scval type"),
+        }
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFromVal<Env, ScAddress> for MuxedAddress {
+    type Error = ConversionError;
+    fn try_from_val(env: &Env, val: &ScAddress) -> Result<Self, Self::Error> {
+        ScVal::Address(val.clone()).try_into_val(env)
+    }
+}
+
+#[cfg(any(test, feature = "testutils"))]
+#[cfg_attr(feature = "docs", doc(cfg(feature = "testutils")))]
+impl crate::testutils::MuxedAddress for MuxedAddress {
+    fn generate(env: &Env, id: u64) -> crate::MuxedAddress {
+        let sc_val = ScVal::Address(crate::env::internal::xdr::ScAddress::MuxedAccount(
+            crate::env::internal::xdr::MuxedEd25519Account {
+                ed25519: crate::env::internal::xdr::Uint256(
+                    env.with_generator(|mut g| g.address()),
+                ),
+                id,
+            },
+        ));
+        sc_val.try_into_val(env).unwrap()
+    }
+
+    fn clone_with_id(&self, new_id: u64) -> crate::MuxedAddress {
```

Comment:
Instead of clone with id, could we add a `new` function with an `Address` and `id`? Then folks could compose the MuxedAddress without the Address having to have already been inside a MuxedAddress. It'll result in more consistent test code that's easier to copy-paste, otherwise the first muxed address has to be created with different code.

Reply:
As I've mentioned above, this would require us adding support for account addresses, and I don't think it's a good idea at the moment. Even if we had that, it would still be confusing, as contract addresses still wouldn't be supported.

> It'll also result in more consistent test code that's easier to copy-paste, 

Yeah, I realize this is not the prettiest interface and I'm open to suggestions on how we could improve it. But for now I think that whatever interface we come up with should not rely on generating account addresses.